publishers/core.py

```diff
@@ -0,0 +1,227 @@
+            new_publication = publisher.publish(alert, new_publication)
+
+        return new_publication
+
+
+class WrappedFunctionPublisher(AlertPublisher):
+    """A class only used to wrap a function publisher."""
+
+    def __init__(self, function):
+        self._function = function
+
+    def publish(self, alert, publication):
+        return self._function(alert, publication)
+
+
+class AlertPublisherRepository(object):
+    """A repository mapping names -> publishers"""
+    _publishers = {}
+
+    @staticmethod
+    def is_valid_publisher(class_or_function):
+        """Returns TRUE if the given reference can be registered as a publisher"""
+        if isclass(class_or_function) and issubclass(class_or_function, AlertPublisher):
+            return True
+        elif callable(class_or_function):
+            return True
+
+        return False
+
+    @staticmethod
+    def get_publisher_name(class_or_function):
+        """Given a class or function, will return its fully qualified name.
+
+            This is useful for assigning a unique string name for a publisher."""
+        return '{}.{}'.format(class_or_function.__module__, class_or_function.__name__)
+
+    @classmethod
+    def register_publisher(cls, publisher):
+        """Registers the publisher into the repository.
+
+        Args:
+             publisher (callable|AlertPublisher): An instance of a publisher class
+
+        Return:
+            void
+        """
+        if not AlertPublisherRepository.is_valid_publisher(publisher):
+            LOGGER.error(
+                'Could not register publisher %s; Not callable nor subclass of AlertPublisher',
+                publisher
+            )
+            return
+
+        # We have to put the isclass() check BEFORE the callable() check because classes are also
+        # callable!
+        elif isclass(publisher):
+            # If the provided publisher is a Class, then we simply need to instantiate an instance
+            # of the class and register it.
+            publisher_instance = publisher()
+        else:
+            # If the provided publisher is a function, we wrap it with a WrappedFunctionPublisher
+            # to make them easier to handle.
+            publisher_instance = WrappedFunctionPublisher(publisher)
+
+        name = AlertPublisherRepository.get_publisher_name(publisher)
+
+        if name in cls._publishers:
+            LOGGER.error('Publisher with name [%s] has already been registered.', name)
+            return
+
+        cls._publishers[name] = publisher_instance
+
+    @classmethod
+    def get_publisher(cls, name):
+        """Returns the subclass that should handle this particular service
+
+        Args:
+            name (str): The name of the publisher.
+
+        Returns:
+            AlertPublisher|None
+        """
+        if cls.has_publisher(name):
+            return cls._publishers[name]
+
+        LOGGER.error('Publisher [%s] does not exist', name)
```

Comment:
This was (kind of..) in the name of compatibility in certain corner cases: 

## Case #1: Accidentally deploy rule without deploying alert
* Develop a new Publisher named `abc`
* `deploy --function rule`
* Since you forgot to deploy the publisher code to the `alert` lambda, alerts get stuck in *DynamoDB* and error like crazy until a deploy occurs.


## Case #2: Removing a publisher
* Suppose you wanted to remove a publisher, `xyz`
* Delete it and `deploy --function alert`
* If there were alerts in flight, queued up on *DynamoDB* now they get permanently stuck unless you modify the row.


In the current implementation, the above alerts will each still send *one* error but still get dispatched regardless, leveraging a default publisher implementation.